cowbird/handlers/impl/geoserver.py

```diff
@@ -126,11 +134,120 @@ def user_deleted(self, user_name):
         Monitoring().unregister(self._shapefile_folder_dir(user_name), self)
 
+    def get_shapefile_list(self, workspace_name, shapefile_name):
+        # type:(str, str) -> List[str]
+        """
+        Generates the list of all files associated with a shapefile name.
+        """
+        base_filename = self._shapefile_folder_dir(workspace_name) + "/" + shapefile_name
+        return [base_filename + ext for ext in SHAPEFILE_ALL_EXTENSIONS]
+
+    def _update_resource_files_permissions(self, resource_type, permission, resource_id, workspace_name,
+                                           layer_name=None):
+        # type:(str, Permission, int, str, Optional[str]) -> None
+        """
+        Updates a Magpie resource's associated files according to its permissions found on Magpie.
+        """
+        if resource_type == "layer":
+            if not layer_name:
+                raise GeoserverError("Missing layer name to update permissions.")
+            file_list = self.get_shapefile_list(workspace_name, layer_name)
+        else:
+            file_list = [self._shapefile_folder_dir(workspace_name)]
+
+        # Get the actual effective user permissions
+        user_permissions = HandlerFactory().get_handler("Magpie").get_user_permissions_by_res_id(
+            permission.user, resource_id, effective=True)
+
+        allowed_user_perm_names = {p["name"] for p in user_permissions["permissions"] if p["access"] == "allow"}
+        is_readable = any(p in LAYER_READ_PERMISSIONS for p in allowed_user_perm_names)
+        is_writable = any(p in LAYER_WRITE_PERMISSIONS for p in allowed_user_perm_names)
+
+        for file in file_list:
+            if not os.path.exists(file):
+                LOGGER.warning("%s could not be found and its permissions could not be updated.", file)
+                continue
+            new_perms = os.stat(file)[stat.ST_MODE]
+            new_perms = update_permissions(new_perms,
+                                           is_readable=is_readable,
+                                           is_writable=is_writable,
+                                           is_executable=is_readable)
+            try:
+                os.chmod(file, new_perms)
+            except PermissionError as exc:
+                LOGGER.warning("Failed to change permissions on the %s file: %s", file, exc)
+            try:
+                # This operation only works as root.
+                os.chown(file, DEFAULT_UID, DEFAULT_GID)
+            except PermissionError as exc:
+                LOGGER.warning("Failed to change ownership of the %s file: %s", file, exc)
+
+    def _update_res_children_files_permissions(self, children_res_tree, permission, workspace_name, layer_name=None):
+        # type:(List, Permission, str, Optional[str]) -> None
+        """
+        Recursive method to update all the file permissions of a resource and its children resources as found on Magpie.
+        """
```

Comment:
I think it should work indeed since we use `effective=True`, but i only tested the `deny` case for the shapefile permission. I will add a case for the `recursive` type.